pkg/substitution/substitution.go

```diff
@@ -118,10 +118,21 @@ func matchGroups(matches []string, pattern *regexp.Regexp) map[string]string {
 }
 
 func ApplyReplacements(in string, replacements map[string]string) string {
-	for k, v := range replacements {
-		in = strings.Replace(in, fmt.Sprintf("$(%s)", k), v, -1)
+	// We use a regular expression and ReplaceAll here to ensure we only expand once.
+	// With strings.Replace in a loop, we could expand variables multiple times.
+	// See https://github.com/tektoncd/pipeline/issues/2093
+	re := regexp.MustCompile(`\$\((.+?\))`)
```

Comment:
I'm not quite sure if there's a better character class than `.*` here.

This function accepts any string map key, and previously did find/replace for any type of key...